core/drivers/gic.c

```diff
@@ -152,154 +162,205 @@ void gic_init(vaddr_t gicc_base, vaddr_t gicd_base)
+	assert(!(read32(gd->gicd_base + GICD_IGROUPR(idx)) & mask));
 
 	/* Disable the interrupt */
-	write32(mask, gic.gicd_base + GICD_ICENABLER(idx));
+	write32(mask, gd->gicd_base + GICD_ICENABLER(idx));
 }
 
-uint32_t gic_read_iar(void)
+static uint32_t gic_read_iar(struct gic_data *gd)
 {
-	return read32(gic.gicc_base + GICC_IAR);
+	return read32(gd->gicc_base + GICC_IAR);
 }
 
-void gic_write_eoir(uint32_t eoir)
+static void gic_write_eoir(struct gic_data *gd, uint32_t eoir)
 {
-	write32(eoir, gic.gicc_base + GICC_EOIR);
+	write32(eoir, gd->gicc_base + GICC_EOIR);
 }
 
-bool gic_it_is_enabled(size_t it) {
+static bool gic_it_is_enabled(struct gic_data *gd, size_t it)
+{
 	size_t idx = it / NUM_INTS_PER_REG;
 	uint32_t mask = 1 << (it % NUM_INTS_PER_REG);
-	return !!(read32(gic.gicd_base + GICD_ISENABLER(idx)) & mask);
+	return !!(read32(gd->gicd_base + GICD_ISENABLER(idx)) & mask);
 }
 
-bool gic_it_get_group(size_t it) {
+static bool __maybe_unused gic_it_get_group(struct gic_data *gd, size_t it)
+{
 	size_t idx = it / NUM_INTS_PER_REG;
 	uint32_t mask = 1 << (it % NUM_INTS_PER_REG);
-	return !!(read32(gic.gicd_base + GICD_IGROUPR(idx)) & mask);
+	return !!(read32(gd->gicd_base + GICD_IGROUPR(idx)) & mask);
 }
 
-uint32_t gic_it_get_target(size_t it) {
+static uint32_t __maybe_unused gic_it_get_target(struct gic_data *gd, size_t it)
+{
 	size_t reg_idx = it / NUM_TARGETS_PER_REG;
-	uint32_t target_shift = (it % NUM_TARGETS_PER_REG) * ITARGETSR_FIELD_BITS;
+	uint32_t target_shift = (it % NUM_TARGETS_PER_REG) *
+				ITARGETSR_FIELD_BITS;
 	uint32_t target_mask = ITARGETSR_FIELD_MASK << target_shift;
 	uint32_t target =
-		read32(gic.gicd_base + GICD_ITARGETSR(reg_idx)) & target_mask;
+		read32(gd->gicd_base + GICD_ITARGETSR(reg_idx)) & target_mask;
+
 	target = target >> target_shift;
 	return target;
 }
 
-void gic_dump_state(void)
+void gic_dump_state(struct gic_data *gd)
 {
 	int i;
-	DMSG("GICC_CTLR: 0x%x", read32(gic.gicc_base + GICC_CTLR));
-	DMSG("GICD_CTLR: 0x%x", read32(gic.gicd_base + GICD_CTLR));
 
-	for (i = 0; i < (int)gic.max_it; i++) {
-		if (gic_it_is_enabled(i)) {
+	DMSG("GICC_CTLR: 0x%x", read32(gd->gicc_base + GICC_CTLR));
+	DMSG("GICD_CTLR: 0x%x", read32(gd->gicd_base + GICD_CTLR));
+
+	for (i = 0; i < (int)gd->max_it; i++) {
+		if (gic_it_is_enabled(gd, i)) {
 			DMSG("irq%d: enabled, group:%d, target:%x", i,
-				gic_it_get_group(i), gic_it_get_target(i));
+			     gic_it_get_group(gd, i), gic_it_get_target(gd, i));
 		}
 	}
 }
+
+void gic_it_handle(struct gic_data *gd)
+{
+	uint32_t iar;
+	uint32_t id;
+
+	iar = gic_read_iar(gd);
+	id = iar & GICC_IAR_IT_ID_MASK;
+
+	if (id == GIC_SPURIOUS_ID)
+		DMSG("ignoring spurious interrupt");
+	else
+		itr_handle(id);
+
+	gic_write_eoir(gd, iar);
+}
+
+static void gic_op_add(struct itr_chip *chip, size_t it,
+		       uint32_t flags __unused)
+{
+	struct gic_data *gd = container_of(chip, struct gic_data, chip);
+
+	gic_it_add(gd, it);
+	gic_it_set_cpu_mask(gd, it, 0xff);
```

Comment:
0xff means any cpu online, while the old evaluated to this cpu only. I'll add a comment.